### ISSUE TITLE
PropOr typings are not safe

### ISSUE THREAD

**Issue by peterreisz** [selfrefactor/rambda#559]
I gathered several cases with the results, please review them and If you agreed fix the typings according to.

```ts
interface FooBar {
  readonly required: string;
  readonly optional?: string;
}
const requiredFoobar: FooBar = {} as any;
const optionalFoobar: undefined | FooBar = {} as any; // Do not define object here, because TS use that instead of the type definitions

const returnsString = propOr('required', 'required', requiredFoobar);
const shouldNotCompile = propOr(undefined, 'required', requiredFoobar); // because the property is required, and the input object is present
const returnsString = propOr('optional', 'optional', requiredFoobar); // when we define the default value it must return with a T type
const returnsStringOrUndefined = propOr(undefined, 'optional', requiredFoobar); // when the output can be undefined than it must return a type T | undefined

const returnsString = propOr('required', 'required', optionalFoobar);
const returnsStringOrUndefined = propOr(undefined, 'required', optionalFoobar);
const returnsString = propOr('optional', 'optional', optionalFoobar);
const returnsStringOrUndefined = propOr(undefined, 'optional', optionalFoobar);

// Should not compile in case of type mismatch
const shouldNotCompile = propOr(1, 'required', requiredFoobar);
const shouldNotCompile = propOr(1, 'optional', requiredFoobar);
const shouldNotCompile = propOr(1, 'required', optionalFoobar);
const shouldNotCompile = propOr(1, 'optional', optionalFoobar);

// Should not compile in case of key is not in the object
const shouldNotCompile = propOr('', 'foo', requiredFoobar);
const shouldNotCompile = propOr('', 'foo', optionalFoobar);
```

I also tried to fix the type definition, it is the best I could do:
```ts
type PropOrValue<T, K extends keyof NonNullable<T>> = T extends NonNullable<T> ? NonNullable<T>[K] : NonNullable<T>[K] | undefined

/**
 * It returns either `defaultValue` or the value of `property` in `obj`.
 */
export function propOr<T, K extends keyof NonNullable<T>>(defaultValue: NonNullable<NonNullable<T>[K]>, property: K, obj: T): NonNullable<NonNullable<T>[K]>;
export function propOr<T, K extends keyof NonNullable<T>>(defaultValue: PropOrValue<T, K>, property: K, obj: T): PropOrValue<T, K>;

export function propOr<T, K extends keyof NonNullable<T>>(defaultValue: NonNullable<NonNullable<T>[K]>, property: K): (obj: T) => NonNullable<NonNullable<T>[K]>;
export function propOr<T, K extends keyof NonNullable<T>>(defaultValue: PropOrValue<T, K>, property: K): (obj: T) => PropOrValue<T, K>;

export function propOr<T, K extends keyof NonNullable<T>>(defaultValue: NonNullable<NonNullable<T>[K]>): FunctionToolbelt.Curry<(property: K, obj: T) => NonNullable<NonNullable<T>[K]>>;
export function propOr<T, K extends keyof NonNullable<T>>(defaultValue: PropOrValue<T, K>): FunctionToolbelt.Curry<(property: K, obj: T) => PropOrValue<T, K>>;

```

**Comment by peterreisz** [comment]
The curried version is not working, and the problem is the same with the propEq. I'll try fix them both and give you more test cases.

**Comment by selfrefactor** [comment]
Much obliged

**Comment by selfrefactor** [comment]
I just applied your suggestion @peterreisz , but curried typings tests doesn't seem to work. Would you take a look at it?

I created a branch `issue-559` with your suggestion and with the failing test, so you can see the failing tests.

**Comment by peterreisz** [comment]
Still thinking an a good solution, but I'm out of ideas. The best version for `propOr` and `propEq` I managed to get so far is this:

```ts
export function propEq<P extends string, T>(propToFind: P, valueToMatch: T, obj: Partial<Record<P, T>> | undefined): boolean;
export function propEq<P extends string, T>(propToFind: P, valueToMatch: T): (obj: Partial<Record<P, T>> | undefined) => boolean;
export function propEq<P extends string>(propToFind: P): {
  <T>(valueToMatch: T, obj: Partial<Record<P, T>> | undefined): boolean;
  <T>(valueToMatch: T): (obj: Partial<Record<P, T>> | undefined) => boolean;
};

export function propOr<T, P extends string>(defaultValue: T, property: P, obj: Partial<Record<P, T>> | undefined): T;
export function propOr<T, P extends string>(defaultValue: T, property: P): (obj: Partial<Record<P, T>> | undefined) => T;
export function propOr<T>(defaultValue: T): {
  <P extends string>(property: P, obj: Partial<Record<P, T>> | undefined): T;
  <P extends string>(property: P): (obj: Partial<Record<P, T>> | undefined) => T;
}
```

Problems with the version above:
- cannot use `undefined` in the valueToMatch or defaultValue without manually specifying the generic types
- cannot use together with other rambda functions without defining the generic types. For example `anyPass` with multiple `propEq` pointing to different props would not compile.

If these pass the tests you can merge and close the PR. In case of I managed to find a better solution I'll open another one ticket. Thanks.

**Comment by selfrefactor** [comment]
So `R.propOr` definitions pass the tests and will be released with the next version - `6.6.0`. The issue will remain open until the release is made. `R.propEq` did not pass the test, so for now this suggestion cannot be applied. Thank you for you effort, as this issue is much deeper than just these 2 methods.

**Comment by selfrefactor** [comment]
Closing this issue as the promised changes are applied with the release of version `6.5.0`.
